projects/hathor-wallet-mobile/0003-token-swap.md

```diff
@@ -0,0 +1,98 @@
+# Summary
+[summary]: #summary
+
+Token swap with Dozer protocol on the mobile wallet.
+
+# Motivation
+[motivation]: #motivation
+
+Trading tokens is probably the main activity users do on DeFi nowadays. And Dozer Finance is the main venue to perform that on Hathor Network.
+
+We want to encourage users to discover new tokens and swap them, so adding this to the mobile wallet, the main wallet used, is a big push in this direction.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Dozer token swap protocol
+
+The Dozer protocol for token swap is implemented as a nano contract on Hathor. 
+The methods being called for a token swap depend on the user input and the exchange expected path returned by the contract.
+
+Once both tokens are chosen the user will enter either the input token value or expected output token value, depending on which input area the user chooses the methods called from the contract will be different but should serve the same overall purpose.
+
+With the tokens and value we can call a method on the contract to calculate the swap information, which returns a tuple with:
+- `path`: Comma-separated string of pool keys to traverse in the format `{tokenA}/{tokenB}/{fee}`
+- `amounts`: List of expected amounts at each step
+- `amount_out`/`amount_in`: Final expected output amount or expected deposit
+- `price_impact`: Overall price impact
+
+The swap information can be used so the user can review the swap and proceed to call the actual swap method.
+
+There are 2 swap methods depending on the number of itens on the calculated `path` if there is only 1 swap or multiple swaps.
+Both methods work similarly by choosing actions to deposit the input tokens and withdraw the output tokens.
+
+### Slippage
+
+The actual exchange rate of a token swap is defined during the swap, so when choosing the input token value you must withdraw less than the expected output amount to account for slippage, if you try to withdraw more than the actual amount the swap will fail, if you withdraw less than the actual amount the exceeding amount will be left as balance for the user on the contract.
+
+The same should apply when the user chooses the output token value, he should deposit more than the expected value to account for slippage and any outstanding amount will be left as balance on the contract.

```

Comment:
question: By using these methods the user can create outstanding balances inside the contract. At a first glance it seems like this could be an attack entrypoint of some kind, as there are two separate tokens per transaction that the contract is in debt to external users.

I can imagine the internal contract state clogging up with data from those balances or an edge case compromising the liquidity because of miscalculations about those outstanding balances.

Maybe the contract code itself already has a way to solve this potential issue, but I think it's important enough to bring it up here, and maybe add a link to the contract source code inside this RFC.